File: src/actions/mark-task-success.ts
import { ChainableCommander, Redis } from 'ioredis';
import { EventType } from '../domain/events/event-type';
import { serializeEvent } from '../domain/events/serialize-event';
import { Task } from '../domain/tasks/task';
import { TaskStatus } from '../domain/tasks/task-status';
import {
  getProcessingListKey,
  getQueueTaskCompleteChannel,
  getQueueTaskSuccessChannel,
  getStallingHashKey,
  getSuccessListKey,
  getTaskKey,
} from '../utils/keys';
import { exec } from '../utils/redis';
import { persistTaskMulti } from './persist-task';

/**
 * @ignore
 */
export const markTaskSuccessMulti = ({
  task,
  queue,
  multi,
  result,
  asOf,
  remove,
}: {
  task: Task;
  queue: string;
  multi: ChainableCommander;
  result?: any;
  asOf: Date;
  remove?: boolean;
}): Task => {
  const taskKey = getTaskKey({ taskId: task.id, queue });
  const processingListKey = getProcessingListKey({ queue });
  const successfulTask: Task = {
    ...task,
    processingEndedAt: asOf,
    status: TaskStatus.Success,
    result,
  };
  if (remove) {
    multi.del(taskKey);
  } else {
    persistTaskMulti({ task: successfulTask, queue, multi });
    multi.lpush(getSuccessListKey({ queue }), task.id);
  }
  multi.lrem(processingListKey, 1, task.id);
  multi.hdel(getStallingHashKey({ queue }), task.id);
  multi.publish(
    getQueueTaskSuccessChannel({ queue }),
    serializeEvent({
      createdAt: new Date(),
      type: EventType.TaskSuccess,
      task: successfulTask,
    }),
  );
  multi.publish(
    getQueueTaskCompleteChannel({ queue }),
    serializeEvent({
      createdAt: new Date(),
      type: EventType.TaskComplete,
      task: successfulTask,
    }),
  );
  return successfulTask;
};

/**
 * @ignore
 */
export const markTaskSuccess = async ({
  task,
  queue,
  client,
  result,
  asOf,
  remove,
}: {
  task: Task;
  queue: string;
  client: Redis;
  result?: any;
  asOf: Date;
  remove?: boolean;
}): Promise<Task> => {
  const multi = client.multi();
  const successfulTask = await markTaskSuccessMulti({
    task,
    queue,
    multi,
    result,
    asOf,
    remove,
  });
  await exec(multi);
  return successfulTask;
};

File: src/lua/mark-task-success.lua
local taskId = KEYS[1]
local taskStatus = KEYS[2]
local resultString = KEYS[3]
local taskKey = KEYS[4]
local asOf = KEYS[5]
local remove = KEYS[6]
local successListKey = KEYS[7]
local processingListKey = KEYS[8]
local stallingHashKey = KEYS[9]
local successEventType = KEYS[10]
local completeEventType = KEYS[11]
local taskSuccessChannel = KEYS[12]
local taskCompleteChannel = KEYS[13]

redis.call('lrem', processingListKey, 1, taskId)
redis.call('hdel', stallingHashKey, taskId)

local taskString = redis.call('get', taskKey)
local taskJson = cjson.decode(taskString)
taskJson['result'] = cjson.decode(resultString)
taskJson['status'] = taskStatus
taskJson['processingEndedAt'] = asOf

if remove == 'true' then
    redis.call('del', taskKey)
else
    redis.call('set', taskKey, cjson.encode(taskJson))
    redis.call('lpush', successListKey, taskId)
end

local successEvent = {
    createdAt = asOf,
    type = successEventType,
    task = taskJson
}
redis.call('publish', taskSuccessChannel, cjson.encode(successEvent))

local completeEvent = {
    createdAt = asOf,
    type = completeEventType,
    task = taskJson
}
redis.call('publish', taskCompleteChannel, cjson.encode(completeEvent))

return cjson.encode(taskJson)
